File: python/packages/spore_eval/src/spore_eval/eval/stats.py
"""Native statistics for the EvalHarness — NO specialized stats lib (Rules
26-28). Mirrors ``rust/crates/spore-eval/src/stats.rs`` byte-for-byte.

Ships:
  * :class:`MetricStats` aggregation (mean, stddev, p50, p95, n).
  * :func:`welch_t_test` — unequal-variance two-sample t with
    Welch-Satterthwaite df and a two-sided p-value via the regularized
    incomplete beta function (Lentz continued fraction). Rule 26.
  * :func:`bootstrap_ci` — percentile confidence interval, default 1000
    iters, using an inline :class:`SplitMix64` PRNG so cross-language replay
    is byte-identical. Rule 27.
"""

from __future__ import annotations

import math
from dataclasses import dataclass

# SplitMix64 / wrapping arithmetic operate on unsigned 64-bit values.
_U64_MASK = 0xFFFF_FFFF_FFFF_FFFF


# ============================================================================
# MetricStats
# ============================================================================


@dataclass
class MetricStats:
    """Aggregated sample statistics for one metric over a set of runs
    (Rule 19)."""

    mean: float
    stddev: float
    p50: float
    p95: float
    n: int

    @classmethod
    def from_samples(cls, samples: list[float]) -> MetricStats:
        """Aggregate samples. ``stddev`` is the sample standard deviation
        (Bessel's correction, n-1); with fewer than two samples it is ``0.0``.
        Percentiles use the nearest-rank method on the sorted samples."""
        n = len(samples)
        if n == 0:
            return cls(mean=0.0, stddev=0.0, p50=0.0, p95=0.0, n=0)
        mean = sum(samples) / n
        if n < 2:
            stddev = 0.0
        else:
            var = sum((x - mean) ** 2 for x in samples) / (n - 1)
            stddev = math.sqrt(var)
        ordered = sorted(samples)
        return cls(
            mean=mean,
            stddev=stddev,
            p50=percentile(ordered, 50.0),
            p95=percentile(ordered, 95.0),
            n=n,
        )


def percentile(ordered: list[float], p: float) -> float:
    """Nearest-rank percentile on already-sorted data. ``p`` is in ``[0, 100]``."""
    if not ordered:
        return 0.0
    if len(ordered) == 1:
        return ordered[0]
    rank = math.ceil((p / 100.0) * len(ordered))
    idx = min(max(rank, 1), len(ordered)) - 1
    return ordered[idx]


# ============================================================================
# Welch's t-test (Rule 26)
# ============================================================================


@dataclass
class WelchResult:
    """Result of a two-sample Welch t-test."""

    t: float
    df: float
    p_value: float


def welch_t_test(a: list[float], b: list[float]) -> WelchResult:
    """Welch's unequal-variance two-sample t-test with Welch-Satterthwaite df
    and a two-sided p-value (Rule 26).

    Degenerate inputs (either sample with n < 2, or both variances zero) return
    ``t = 0``, ``df = 0``, ``p_value = 1.0`` — "no detectable difference" rather
    than a divide-by-zero."""
    na = len(a)
    nb = len(b)
    if na < 2 or nb < 2:
        return WelchResult(t=0.0, df=0.0, p_value=1.0)
    mean_a = sum(a) / na
    mean_b = sum(b) / nb
    var_a = sum((x - mean_a) ** 2 for x in a) / (na - 1)
    var_b = sum((x - mean_b) ** 2 for x in b) / (nb - 1)

    sa = var_a / na
    sb = var_b / nb
    denom = sa + sb
    if denom <= 0.0:
        # Both samples are constant.
        p = 1.0 if abs(mean_a - mean_b) <= 2.220446049250313e-16 else 0.0
        return WelchResult(t=0.0, df=0.0, p_value=p)

    t = (mean_a - mean_b) / math.sqrt(denom)
    df = denom**2 / (sa**2 / (na - 1) + sb**2 / (nb - 1))
    p_value = _two_sided_p(t, df)
    return WelchResult(t=t, df=df, p_value=p_value)


def _two_sided_p(t: float, df: float) -> float:
    """Two-sided p-value for a t-statistic with ``df`` degrees of freedom:
    ``p = I_{df/(df+t^2)}(df/2, 1/2)``."""
    if df <= 0.0:
        return 1.0
    x = df / (df + t * t)
    return min(max(_betai(df / 2.0, 0.5, x), 0.0), 1.0)


def _betai(a: float, b: float, x: float) -> float:
    """Regularized incomplete beta function ``I_x(a, b)`` via the Lentz
    continued fraction (Numerical Recipes' ``betai``)."""
    if x <= 0.0:
        return 0.0
    if x >= 1.0:
        return 1.0
    ln_beta = _ln_gamma(a + b) - _ln_gamma(a) - _ln_gamma(b)
    front = math.exp(a * math.log(x) + b * math.log(1.0 - x) + ln_beta)
    if x < (a + 1.0) / (a + b + 2.0):
        return front * _betacf(a, b, x) / a
    return 1.0 - front * _betacf(b, a, 1.0 - x) / b


def _betacf(a: float, b: float, x: float) -> float:
    """Lentz continued fraction for the incomplete beta."""
    max_iter = 200
    eps = 3.0e-12
    fpmin = 1.0e-300

    qab = a + b
    qap = a + 1.0
    qam = a - 1.0
    c = 1.0
    d = 1.0 - qab * x / qap
    if abs(d) < fpmin:
        d = fpmin
    d = 1.0 / d
    h = d

    for m_int in range(1, max_iter + 1):
        m = float(m_int)
        m2 = 2.0 * m
        # Even step.
        aa = m * (b - m) * x / ((qam + m2) * (a + m2))
        d = 1.0 + aa * d
        if abs(d) < fpmin:
            d = fpmin
        c = 1.0 + aa / c
        if abs(c) < fpmin:
            c = fpmin
        d = 1.0 / d
        h *= d * c
        # Odd step.
        aa = -(a + m) * (qab + m) * x / ((a + m2) * (qap + m2))
        d = 1.0 + aa * d
        if abs(d) < fpmin:
            d = fpmin
        c = 1.0 + aa / c
        if abs(c) < fpmin:
            c = fpmin
        d = 1.0 / d
        delta = d * c
        h *= delta
        if abs(delta - 1.0) < eps:
            break
    return h


_LANCZOS_G = 7.0
_LANCZOS_COEF = (
    0.9999999999998099,
    676.5203681218851,
    -1259.1392167224028,
    771.3232867877531,
    -176.6150291621406,
    12.507343278686905,
    -0.13857109526572012,
    9.984369578019572e-6,
    1.5056327351493116e-7,
)


def _ln_gamma(x: float) -> float:
    """Lanczos approximation of ``ln(gamma(x))`` for x > 0."""
    if x < 0.5:
        # Reflection formula.
        return math.log(math.pi) - math.log(math.sin(math.pi * x)) - _ln_gamma(1.0 - x)
    x -= 1.0
    a = _LANCZOS_COEF[0]
    t = x + _LANCZOS_G + 0.5
    for i in range(1, len(_LANCZOS_COEF)):
        a += _LANCZOS_COEF[i] / (x + i)
    return 0.5 * math.log(2.0 * math.pi) + (x + 0.5) * math.log(t) - t + math.log(a)


# ============================================================================
# Bootstrap CI (Rule 27)
# ============================================================================


@dataclass
class ConfidenceInterval:
    """A bootstrap confidence interval."""

    lower: float
    upper: float
    level: float


class SplitMix64:
    """Inline SplitMix64 PRNG (Rule 27): a fixed, seedable, byte-identical
    generator so a bootstrap CI replays identically across all four
    languages."""

    __slots__ = ("state",)

    def __init__(self, seed: int) -> None:
        self.state = seed & _U64_MASK

    def next_u64(self) -> int:
        """Next 64-bit value (the canonical SplitMix64 step)."""
        self.state = (self.state + 0x9E37_79B9_7F4A_7C15) & _U64_MASK
        z = self.state
        z = ((z ^ (z >> 30)) * 0xBF58_476D_1CE4_E5B9) & _U64_MASK
        z = ((z ^ (z >> 27)) * 0x94D0_49BB_1331_11EB) & _U64_MASK
        return z ^ (z >> 31)

    def next_index(self, n: int) -> int:
        """Uniform index in ``[0, n)``. ``n`` must be non-zero."""
        return self.next_u64() % n


def bootstrap_ci(
    samples: list[float],
    iterations: int,
    level: float,
    seed: int,
) -> ConfidenceInterval | None:
    """Percentile bootstrap CI for the mean (Rule 27). Resamples ``samples``
    with replacement ``iterations`` times using a :class:`SplitMix64` seeded
    with ``seed``, then takes the ``[(1-level)/2, 1-(1-level)/2]`` percentiles
    of the resample means. Returns ``None`` for an empty sample."""
    if not samples:
        return None
    rng = SplitMix64(seed)
    n = len(samples)
    means: list[float] = []
    for _ in range(iterations):
        total = 0.0
        for _ in range(n):
            total += samples[rng.next_index(n)]
        means.append(total / n)
    means.sort()
    alpha = 1.0 - level
    lower = percentile(means, alpha / 2.0 * 100.0)
    upper = percentile(means, (1.0 - alpha / 2.0) * 100.0)
    return ConfidenceInterval(lower=lower, upper=upper, level=level)


DEFAULT_BOOTSTRAP_ITERATIONS = 1000
"""Default bootstrap iteration count (Rule 27)."""

DEFAULT_BOOTSTRAP_SEED = 0x5EED_5EED_5EED_5EED
"""Fixed bootstrap seed so replays are byte-identical cross-language."""

File: typescript/packages/eval/src/stats.ts
/**
 * Native statistics for the EvalHarness — NO external stats library (Rules 26-28).
 *
 * Ships:
 *   - {@link MetricStats} aggregation (mean, stddev, p50, p95, n).
 *   - {@link welchTTest} — unequal-variance two-sample t with Welch–Satterthwaite
 *     df and a two-sided p-value via the regularized incomplete beta function
 *     (Lentz continued fraction). Rule 26.
 *   - {@link bootstrapCi} — percentile confidence interval, default 1000 iters,
 *     using an inline SplitMix64 PRNG so cross-language replay is byte-identical.
 *     Rule 27.
 *
 * Every public function has a hand-computed oracle test (Rule 28); the seeded
 * bootstrap reproduces the `welch_bootstrap.json` oracle byte-for-byte (Rule 29).
 */

// ============================================================================
// MetricStats
// ============================================================================

/** Aggregated sample statistics for one metric over a set of runs (Rule 19). */
export interface MetricStats {
  mean: number;
  stddev: number;
  p50: number;
  p95: number;
  n: number;
}

/**
 * Aggregate a slice of samples. `stddev` is the **sample** standard deviation
 * (Bessel's correction, n-1); with fewer than two samples it is `0`.
 * Percentiles use the nearest-rank method on the sorted samples.
 */
export function metricStatsFromSamples(
  samples: readonly number[],
): MetricStats {
  const n = samples.length;
  if (n === 0) {
    return { mean: 0, stddev: 0, p50: 0, p95: 0, n: 0 };
  }
  const mean = samples.reduce((acc, x) => acc + x, 0) / n;
  let stddev = 0;
  if (n >= 2) {
    const variance =
      samples.reduce((acc, x) => acc + (x - mean) ** 2, 0) / (n - 1);
    stddev = Math.sqrt(variance);
  }
  const sorted = [...samples].sort((a, b) => a - b);
  return {
    mean,
    stddev,
    p50: percentile(sorted, 50),
    p95: percentile(sorted, 95),
    n,
  };
}

/** Nearest-rank percentile on already-sorted data. `p` is in `[0, 100]`. */
export function percentile(sorted: readonly number[], p: number): number {
  if (sorted.length === 0) return 0;
  if (sorted.length === 1) return sorted[0]!;
  // Nearest-rank: rank = ceil(p/100 * n), clamped to [1, n].
  const rank = Math.ceil((p / 100) * sorted.length);
  const idx = clamp(rank, 1, sorted.length) - 1;
  return sorted[idx]!;
}

function clamp(value: number, lo: number, hi: number): number {
  return Math.min(Math.max(value, lo), hi);
}

// ============================================================================
// Welch's t-test (Rule 26)
// ============================================================================

/** Result of a two-sample Welch t-test. */
export interface WelchResult {
  t: number;
  df: number;
  /** Two-sided p-value. */
  pValue: number;
}

/**
 * Welch's unequal-variance two-sample t-test with Welch–Satterthwaite degrees
 * of freedom and a two-sided p-value (Rule 26).
 *
 * Degenerate inputs (either sample with n < 2, or both variances zero) return
 * `t = 0`, `df = 0`, `pValue = 1` — "no detectable difference" rather than a
 * divide-by-zero.
 */
export function welchTTest(
  a: readonly number[],
  b: readonly number[],
): WelchResult {
  const na = a.length;
  const nb = b.length;
  if (na < 2 || nb < 2) {
    return { t: 0, df: 0, pValue: 1 };
  }
  const meanA = a.reduce((acc, x) => acc + x, 0) / na;
  const meanB = b.reduce((acc, x) => acc + x, 0) / nb;
  const varA = a.reduce((acc, x) => acc + (x - meanA) ** 2, 0) / (na - 1);
  const varB = b.reduce((acc, x) => acc + (x - meanB) ** 2, 0) / (nb - 1);

  const sa = varA / na;
  const sb = varB / nb;
  const denom = sa + sb;
  if (denom <= 0) {
    // Both samples are constant. Equal means → no difference; differing
    // constant means → an infinitely significant difference (p → 0).
    const p = Math.abs(meanA - meanB) <= Number.EPSILON ? 1 : 0;
    return { t: 0, df: 0, pValue: p };
  }

  const t = (meanA - meanB) / Math.sqrt(denom);
  // Welch–Satterthwaite df.
  const df = denom ** 2 / (sa ** 2 / (na - 1) + sb ** 2 / (nb - 1));
  const pValue = twoSidedP(t, df);
  return { t, df, pValue };
}

/**
 * Two-sided p-value for a t-statistic with `df` degrees of freedom, via the
 * regularized incomplete beta function:
 *   p = I_{df/(df+t^2)}(df/2, 1/2)
 */
function twoSidedP(t: number, df: number): number {
  if (df <= 0) return 1;
  const x = df / (df + t * t);
  return clamp(betai(df / 2, 0.5, x), 0, 1);
}

/**
 * Regularized incomplete beta function I_x(a, b) via the Lentz continued
 * fraction, the same routine used by Numerical Recipes' `betai`.
 */
function betai(a: number, b: number, x: number): number {
  if (x <= 0) return 0;
  if (x >= 1) return 1;
  const lnBeta = lnGamma(a + b) - lnGamma(a) - lnGamma(b);
  const front = Math.exp(a * Math.log(x) + b * Math.log(1 - x) + lnBeta);
  // Use the symmetry relation for faster continued-fraction convergence.
  if (x < (a + 1) / (a + b + 2)) {
    return (front * betacf(a, b, x)) / a;
  }
  return 1 - (front * betacf(b, a, 1 - x)) / b;
}

/** Lentz continued fraction for the incomplete beta. */
function betacf(a: number, b: number, x: number): number {
  const MAX_ITER = 200;
  const EPS = 3.0e-12;
  const FPMIN = 1.0e-300;

  const qab = a + b;
  const qap = a + 1;
  const qam = a - 1;
  let c = 1;
  let d = 1 - (qab * x) / qap;
  if (Math.abs(d) < FPMIN) d = FPMIN;
  d = 1 / d;
  let h = d;

  for (let m = 1; m <= MAX_ITER; m++) {
    const m2 = 2 * m;
    // Even step.
    let aa = (m * (b - m) * x) / ((qam + m2) * (a + m2));
    d = 1 + aa * d;
    if (Math.abs(d) < FPMIN) d = FPMIN;
    c = 1 + aa / c;
    if (Math.abs(c) < FPMIN) c = FPMIN;
    d = 1 / d;
    h *= d * c;
    // Odd step.
    aa = (-(a + m) * (qab + m) * x) / ((a + m2) * (qap + m2));
    d = 1 + aa * d;
    if (Math.abs(d) < FPMIN) d = FPMIN;
    c = 1 + aa / c;
    if (Math.abs(c) < FPMIN) c = FPMIN;
    d = 1 / d;
    const del = d * c;
    h *= del;
    if (Math.abs(del - 1) < EPS) break;
  }
  return h;
}

const LN_GAMMA_G = 7;
const LN_GAMMA_COEF = [
  0.9999999999998099, 676.5203681218851, -1259.1392167224028, 771.3234287776531,
  -176.6150291621406, 12.507343278686905, -0.13857109526572012,
  9.984369578019572e-6, 1.5056327351493116e-7,
] as const;

/** Lanczos approximation of ln(Γ(x)) for x > 0. */
function lnGamma(x: number): number {
  if (x < 0.5) {
    // Reflection formula.
    return Math.log(Math.PI) - Math.log(Math.sin(Math.PI * x)) - lnGamma(1 - x);
  }
  const z = x - 1;
  let a = LN_GAMMA_COEF[0]!;
  const t = z + LN_GAMMA_G + 0.5;
  for (let i = 1; i < LN_GAMMA_COEF.length; i++) {
    a += LN_GAMMA_COEF[i]! / (z + i);
  }
  return (
    0.5 * Math.log(2 * Math.PI) + (z + 0.5) * Math.log(t) - t + Math.log(a)
  );
}

// ============================================================================
// Bootstrap CI (Rule 27)
// ============================================================================

/**
 * A bootstrap confidence interval. Mirrors the `ConfidenceInterval` reported on
 * a {@link import("./report.js").MetricComparison}.
 */
export interface ConfidenceInterval {
  lower: number;
  upper: number;
  level: number;
}

/**
 * Inline SplitMix64 PRNG (Rule 27): a fixed, seedable, byte-identical generator
 * so a bootstrap CI replays identically across all four languages.
 *
 * Implemented over `BigInt` u64 arithmetic so the wrapping multiplies match the
 * Rust reference exactly (JS `number` cannot hold 64-bit integer products).
 */
export class SplitMix64 {
  private state: bigint;

  constructor(seed: bigint) {
    this.state = BigInt.asUintN(64, seed);
  }

  /** Next 64-bit value (the canonical SplitMix64 step). */
  nextU64(): bigint {
    this.state = BigInt.asUintN(64, this.state + 0x9e3779b97f4a7c15n);
    let z = this.state;
    z = BigInt.asUintN(64, (z ^ (z >> 30n)) * 0xbf58476d1ce4e5b9n);
    z = BigInt.asUintN(64, (z ^ (z >> 27n)) * 0x94d049bb133111ebn);
    return z ^ (z >> 31n);
  }

  /** Uniform index in `[0, n)`. `n` must be a positive integer. */
  nextIndex(n: number): number {
    return Number(this.nextU64() % BigInt(n));
  }
}

/** Default bootstrap iteration count (Rule 27). */
export const DEFAULT_BOOTSTRAP_ITERATIONS = 1000;
/** Fixed bootstrap seed so replays are byte-identical cross-language. */
export const DEFAULT_BOOTSTRAP_SEED = 0x5eed5eed5eed5eedn;

/**
 * Percentile bootstrap CI for the mean (Rule 27). Resamples `samples` with
 * replacement `iterations` times using a SplitMix64 seeded with `seed`, then
 * takes the `[(1-level)/2, 1-(1-level)/2]` percentiles of the resample means.
 *
 * `level` is the confidence level (e.g. `0.95`). Returns `undefined` for an
 * empty sample.
 */
export function bootstrapCi(
  samples: readonly number[],
  iterations: number,
  level: number,
  seed: bigint,
): ConfidenceInterval | undefined {
  if (samples.length === 0) return undefined;
  const rng = new SplitMix64(seed);
  const means: number[] = new Array(iterations);
  for (let i = 0; i < iterations; i++) {
    let sum = 0;
    for (let j = 0; j < samples.length; j++) {
      sum += samples[rng.nextIndex(samples.length)]!;
    }
    means[i] = sum / samples.length;
  }
  means.sort((a, b) => a - b);
  const alpha = 1 - level;
  const lower = percentile(means, (alpha / 2) * 100);
  const upper = percentile(means, (1 - alpha / 2) * 100);
  return { lower, upper, level };
}
